Implements clean output - compress/archive output files

joaomcteixeira opened the pull request:
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and you comply with the following criteria:

- [x] You have stick to Python. Talk with us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there's a (state) purpose
- [x] code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any *install dependencies* unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Closes #448

Meanwhile I write docs and tests, please test the clean output functionality and let me know your feedback. Run any example and you will see the output is compressed by default.

* `seed`, `inp`, `out`, and `con` files are archived (and compressed) in `tar` files.
* `pdb` and `psf` files are compressed in `gz`
* works with `--restart` and `--extend-run` options to the best of my knowledge
* a new parameter is introduced: `clean=true/false`. It works as an optional general parameter. Can be defined generally or specifically for each module.

joaomcteixeira commented:
@rvhonorato I took a look to your suggestion using https://github.com/vinlyx/mgzip. But the project was untouched for several years now. So, I implemented something simple on my own with `multiprocessing`. When our clusters are back I will test the speed of this in full runs to see the performance of the compressing routines and let your know.

amjjbonvin commented on src/haddock/gear/clean_steps.py:
```diff
@@ -0,0 +1,87 @@
+
+    for folder in folders:
+        gz_files = glob_folder(folder, '.gz')
+        tar_files = glob_folder(folder, '.tar')
```
Shouldn't you rather have `.tgz` as extension if you generate compressed tar archives (giving the `z` option to tar)?

amjjbonvin commented on src/haddock/libs/libio.py:
```diff
@@ -138,3 +145,173 @@ def working_directory(path):
+    Returns
+    -------
+    bool
+        ``True`` if files with ``ext`` were found and the Zip files created.
```
Didn't you dropped `zip` files in favour of `tgz`?

amjjbonvin commented on src/haddock/libs/libio.py:
```diff
@@ -138,3 +145,173 @@ def working_directory(path):
+    Parameters
+    ----------
+    file_ : str or :external:py:class:`pathlib.Path`
+        The path to the file to zip.
```
The path to the file to zip.... Unclear meaning

amjjbonvin commented on src/haddock/modules/defaults.yaml:
```diff
@@ -94,3 +94,14 @@ self_contained:
     main installation.
   group: 'execution'
   explevel: guru
+clean:
```
If the cleaning is not a module that the  user needs to add at the end, it might be safer to make the default to false for the time being. This parameter should be specified in all our examples cfg file, possibly as commented out, but to make clear it can be used.

amjjbonvin commented on src/haddock/modules/defaults.yaml:
```diff
@@ -94,3 +94,14 @@ self_contained:
+    succeeds. In this process, PDB files are compressed to gzip. `.seed`, `.inp`, and `.out`
+    files are archived.
+  group: 'clean'
+  explevel: guru
```
If we switch to false as default, it should be usable at the easy level.

rvhonorato reviewed (changes requested):
On top of what @amjjbonvin said I'm just concerned about performance, could you please run it on a large sampling so we can at least check if its a bottleneck or not?

amjjbonvin commented:
Should not be a bid deal I would say. 

Would be nice to be able to do it at the end of successful run by giving the —restart option 
But for that to work if would have to be a separate module I guess.

Or call haddock3 with a —clean option on a completed run

joaomcteixeira commented:
I was waiting for our servers to be operational again to test it. I am also
curious. I will do it tomorrow.

Yes, having a CLI to clean a successful module is a good idea. I will add
it to this PR. It is also a good example on how seamlessly our core
functions can be used in the clients.

I tried this implementation with '--restart' and '--extend-run' and the
automatic compressing/uncompressing of previous compressed data is working.
haddock3 will register the folder was compress and will compress it back
again once the restart or extend finish.

the "clean" parameter can also be set globally or for each module.

El mié., 8 jun. 2022 12:50, Alexandre Bonvin ***@***.***>
escribió:

> Should not be a bid deal I would say.
>
> Would be nice to be able to do it at the end of successful run by giving
> the —restart option
> But for that to work if would have to be a separate module I guess.
>
> Or call haddock3 with a —clean option on a completed run
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/haddocking/haddock3/pull/468#issuecomment-1149761082>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWEOMR2GGLTWQ5YY3727QLVOB3HBANCNFSM5YA5KTOA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

joaomcteixeira commented:
Addressed comments, finished the implementations, and added docs.

Also tested the performance. On `tintin` with `0_topoaa` and `1_rigidbody` with `sampling=10000` (ten thousand), on a full node took 5 minutes.

```
(haddock3) [-@tintin docking-protein-protein]$ haddock3-clean -r run10000/
[2022-06-12 19:17:07,677 clean_steps INFO] Cleaning output for 'run10000/0_topoaa' using 95 cores.
[2022-06-12 19:17:08,245 libtimer INFO] compressing took 1 seconds
[2022-06-12 19:17:08,245 clean_steps INFO] Cleaning output for 'run10000/1_rigidbody' using 95 cores.
[2022-06-12 19:21:14,388 libtimer INFO] compressing took 4 minutes and 6 seconds
```

To unpack took about 30 seconds:

```
(haddock3) [-@tintin docking-protein-protein]$ haddock3-unpack -r run10000/
[2022-06-12 19:22:55,621 libtimer INFO] unpacking took 26 seconds
```

Is the performance okay? Everything else is done. You can review, we discuss, and we merge.

joaomcteixeira commented:
The last commits:

* further increase the performance to `3 minutes and 37 seconds` to clean a `rigidbody` folder with `sampling=10000` from the `docking-protein-protein` example.
* improves the clients to allow cleaning/unpack specific step folders
* adds documentation

mgiulini reviewed (approved):
good work! as for the tests functions, should we add few words of comment to each one? or are we creating a separate issue for that?

joaomcteixeira commented:
I can add additional comments on the test functions.

mgiulini commented:
> I can add additional comments on the test functions.

super brief, just three words to know that `test_is_step_folder` is checking the correct behavior of is_step_folder.

But you can of course ignore it for now :)
